### PR TITLE
Use lua_tolstring and lua_pushlstring where possible

### LINKED ISSUE #9734: Use lua_tolstring and lua_pushlstring where possible
**Task (REQUIRED):**

`lua_tolstring` and `lua_pushlstring` (those with a length argument) avoids an unnecessary `strlen` call, so should be preferred over the non-length ones.

**Expected outcome (REQUIRED):**

Change the code accordingly, where length is needed or can easily be retrieved.

### PR DESCRIPTION
**Pull request by redoak**
skip release notes

Resolves #9734

### Technical changes

* No changes to actual behavior — simply makes use of string length where it's easy enough to do.

### CODE REVIEW

File: engine/script/src/script_sys.cpp [hunk 1]
```diff
@@ -1201,12 +1201,12 @@ union SaveLoadBuffer
...
-    * @param [arg3] [type:string] argument 3
-    * @param [arg4] [type:string] argument 4
-    * @param [arg5] [type:string] argument 5
-    * @param [arg6] [type:string] argument 6
+    * @param [arg1] [type:string|number] argument 1
+    * @param [arg2] [type:string|number] argument 2
+    * @param [arg3] [type:string|number] argument 3
+    * @param [arg4] [type:string|number] argument 4
+    * @param [arg5] [type:string|number] argument 5
+    * @param [arg6] [type:string|number] argument 6
```
Comment (JCash) [inline engine/script/src/script_sys.cpp:1209]: The interface is a string.
I don't think we should tell users to pass in a number, when we use checkstring.
Also, this change is unrelated to the rest of the PR.
This is what we meant with keeping changes separate.
If we need to revert the PR, _everything_ will be reverted.

Please undo this from this PR.

File: engine/script/src/script_table.cpp [hunk 2]
```diff
@@ -265,7 +265,7 @@ namespace dmScript
             luaL_error(L, "Reading outside of buffer at element #%d (string): wanted to read: %d bytes left: %d [BufStart: %p, BufSize: %lu]\n'%s'", count, total_size, (int)(buffer_end - buffer), logger.m_BufferStart, logger.m_BufferSize, log_str);
         }
 
-        lua_pushstring(L, buffer);
+        lua_pushlstring(L, buffer, total_size - 1);
```
Comment (JCash) [inline engine/script/src/script_table.cpp:268]: This is another case where one should be very cautious.
The load + save functions are volatile: Messing this up will break save games for all games.
While this is a correct fix, it just feels a bit unnecessary to "optimize" it.

**Review by JCash — CHANGES_REQUESTED** [review]

**Comment by JCash** [comment]
Q: What platforms have you run the tests on?

File: engine/gamesys/src/gamesys/scripts/script_buffer.cpp [hunk 3]
```diff
@@ -755,18 +755,18 @@ namespace dmGameSystem
         }
 
         char buf[128];
-        uint32_t maxlen = 64 + num_streams * sizeof(buf);
-        char* s = (char*)alloca( maxlen );
+        uint32_t ssize = 64 + num_streams * sizeof(buf);
```
[the remark this answers is not in the document]
Comment (redoak) [reply engine/gamesys/src/gamesys/scripts/script_buffer.cpp:758]: I wasn't aware it's a thing. Learning something new every day!

File: engine/gamesys/src/gamesys/scripts/script_buffer.cpp [hunk 4]
```diff
@@ -780,11 +780,11 @@ namespace dmGameSystem
...
             const char* typestring = dmBuffer::GetValueTypeString(type);
             dmSnPrintf(buf, sizeof(buf), "{ hash(\"%s\"), buffer.%s, %d }%s", dmHashReverseSafe64(stream_name), typestring, type_count, comma );
-            dmStrlCat(s, buf, maxlen);
+            dmStrlCat(s, buf, ssize);
         }
-        dmStrlCat(s, ")", maxlen);
+        int length = (int)dmStrlCat(s, ")", ssize);
 
-        lua_pushstring(L, s);
+        lua_pushlstring(L, s, length < ssize ? length : ssize - 1);
```
[the remark this answers is not in the document]
Comment (redoak) [reply engine/gamesys/src/gamesys/scripts/script_buffer.cpp:787]: `dmStrlCat` will return the untruncated length, which would result in `lua_pushlstring` reading beyond the end of the buffer.

File: engine/script/src/script_extensions.cpp [hunk 5]
```diff
@@ -29,7 +29,7 @@ namespace dmScript
     #define BIT_INDEX(b) ((b) / sizeof(uint32_t))
     #define BIT_OFFSET(b) ((b) % sizeof(uint32_t))
 
-    const char* EXTENSIONS_CONTEXT_NAME = "__extensions_context__";
+    const char EXTENSIONS_CONTEXT_NAME[] = "__extensions_context__";
```
[the remark this answers is not in the document]
Comment (redoak) [reply engine/script/src/script_extensions.cpp:32]: With this change, `sizeof` will return the size of the array rather than the size of the pointer. This is needed for using `sizeof` together with `lua_pushlstring` below.

File: engine/script/src/script_extensions.cpp [hunk 6]
```diff
@@ -52,7 +52,7 @@ namespace dmScript
         // [-1] extension_data
         extension_data->m_Ref = Ref(L, LUA_REGISTRYINDEX);
         // [-1] extension_data
-        lua_pushstring(L, EXTENSIONS_CONTEXT_NAME);
+        lua_pushlstring(L, EXTENSIONS_CONTEXT_NAME, sizeof(EXTENSIONS_CONTEXT_NAME) - 1);
```
[the remark this answers is not in the document]
Comment (redoak) [reply engine/script/src/script_extensions.cpp:55]: `lua_pushliteral` unfortunately requires the argument to be an inline literal (i.e. "string" or a #defined string).

File: engine/script/src/script_sys.cpp (hunk 1 above)
Comment (redoak) [reply engine/script/src/script_sys.cpp:1209]: I "corrected" the documentation since I touched the function. But I get your point.

I added `number` since `lua_tostring` will convert number to string. Maybe the Defold team has discussed this and how to handle it in code and docs — I wanted to bring it up so I know what's deemed correct.

Personally, I prefer to have API documentation that is true to the behavior of the function, even if it makes the docs less concise. Silently converting number arguments may cause user confusion and issues for tool authors that parse the docs.

If the function should only accept strings, then it should call `luaL_checktype(L, i, LUA_TSTRING)` before `lua_tostring`.

Allowing implicit number-to-string and string-to-number argument conversions is arguably desirable, as it's expected in Lua. But it also creates situations where the user by mistake might pass a number to an engine function, which Lua will convert to a string, and the engine will then hash — that's not an insignificant amount of work to happen implicitly.

If you think this is a worthwhile discussion to have, I can open an issue about it.

**Comment by redoak** [comment]
> Q: What platforms have you run the tests on?

Only Windows. Unfortunately, I'm unable to build web. I could do Android, through.

**Comment by redoak** [comment]
Since most of the changes will either increase build size or are inadvisable as pointed out in the comments, I'm closing this PR as the benefits of what would remain is negligible.

File: engine/script/src/script_msg.cpp [hunk 7]
```diff
@@ -386,13 +386,13 @@ namespace dmScript
...
+                    size_t path_length;
+                    const char* path = lua_tolstring(L, 2, &path_length);
                     if (lua_isnil(L, 1) || (lua_isstring(L, 1) && *lua_tostring(L, 1) == '\0'))
                     {
-                        size_t path_size = strlen(path);
-                        if (path_size > 0)
+                        if (path_length > 0)
                         {
-                            ResolvePath(L, path, path_size, url.m_Path);
+                            ResolvePath(L, path, path_length, url.m_Path);
```
[the remark this answers is not in the document]
Comment (redoak) [reply engine/script/src/script_msg.cpp:395]: When talking about C strings, I'm used to size being the buffer size (i.e. including the space needed for terminating null), and length excluding the null character. When these terms are used interchangeably in the code base, it can be frustrating as the user needs to double-check what value is assigned to it. But it may not be as well-established terms as I've assumed.

In any case, you're right in that I shouldn't pollute the changeset with this kind of thing.
